### PR TITLE
fix: predicate logic against bools

### PR DESCRIPTION
**Pull request by svidgen**
#### Description of changes

Predicate filtering logic (used by `observe` and `observeQuery`) was not working correctly against bools, because we were casting field values to strings before comparison. This appears to be cruft from POC work &mdash; I can't otherwise tell why this cast would have been in place.

#### Description of how you validated changes

Added tests that were red. Made the change. Tests turned green.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/datastore/__tests__/helpers.ts [hunk 1]
```diff
@@ -1478,6 +1482,20 @@ export declare class LegacyJSONComment {
 	): LegacyJSONComment;
 }
 
+export declare class ModelWithBoolean {
+	public readonly id: string;
+	public readonly boolField: boolean;
```
Comment (svidgen) [inline packages/datastore/__tests__/helpers.ts:1487]:
```suggestion
	public readonly boolField?: boolean;
```

File: packages/datastore/__tests__/model.ts [hunk 2]
```diff
@@ -138,6 +138,7 @@ declare class EditorModel {
 
 declare class AuthorModel {
 	readonly id: string;
+	readonly isActive: boolean;
```
Comment (svidgen) [inline packages/datastore/__tests__/model.ts:141]:
```suggestion
	readonly isActive?: boolean;
```

**Review by iartemiev — APPROVED** [review]
LGTM 👍🏻

**Review by dpilch — CHANGES_REQUESTED** [review]
Can you add tests for `eq` and `ne` on number fields. I found that casting to string also affects numbers. https://github.com/dpilch/predicate-tests

File: packages/datastore/__tests__/DataStore.ts [hunk 3]
```diff
@@ -1731,6 +1732,96 @@ describe('Model behavior', () => {
 		expect(disconnectedParent.hasOneParentChildId).toEqual(child.id);
 		expect(await disconnectedParent.child).toBeUndefined();
 	});
+
+	test('removes items update out of the from the snapshot with an eq predicate on boolean field', done => {
```
Comment (stocaaro) [inline packages/datastore/__tests__/DataStore.ts:1736]: The wording on this test  description is hard to parse.

File: packages/datastore/__tests__/DataStore.ts [hunk 4]
```diff
@@ -1731,6 +1732,96 @@ describe('Model behavior', () => {
...
+		(async () => {
+			const { DataStore, ModelWithBoolean } = getDataStore();
+			try {
+				const expecteds = [5, 4];
+
+				for (let i = 0; i < 5; i++) {
```
Comment (stocaaro) [inline packages/datastore/__tests__/DataStore.ts:1742]: Why the 5 iterations here? I suppose we're just making 5 discreet records?

**Review by dpilch — APPROVED** [review]

File: packages/datastore/__tests__/DataStore.ts [hunk 5]
```diff
@@ -1731,6 +1732,96 @@ describe('Model behavior', () => {
...
+						new ModelWithBoolean({
+							boolField: true,
+						})
+					);
+				}
+
+				const sub = DataStore.observeQuery(ModelWithBoolean, m =>
+					m.boolField.eq(true)
+				).subscribe(({ items, isSynced }) => {
+					const expected = expecteds.shift() || 0;
```
Comment (stocaaro) [inline packages/datastore/__tests__/DataStore.ts:1753]: So, we get a sequence of true models where the first one has 5 items and the second one has 4 items, then every event after that has 0 items?

I'm not quite sure how these item counts come out this way.

**Review by iartemiev — APPROVED** [review]

File: packages/datastore/__tests__/DataStore.ts [hunk 6]
```diff
@@ -1731,6 +1732,96 @@ describe('Model behavior', () => {
...
+
+					console.log({ items });
+
+					for (let i = 0; i < expected; i++) {
+						expect(items[i].boolField).toEqual(true);
+					}
+
+					if (expecteds.length === 0) {
+						sub.unsubscribe();
+						done();
```
Comment (stocaaro) [inline packages/datastore/__tests__/DataStore.ts:1764]: Got it, so after the second event arrives, exit the test.

File: packages/datastore/__tests__/DataStore.ts [hunk 7]
```diff
@@ -1731,6 +1732,96 @@ describe('Model behavior', () => {
...
+					})
+				);
+				jest.advanceTimersByTime(2000);
+			} catch (error) {
+				done(error);
+			}
+		})();
+	});
+
+	test('removes items update out of the from the snapshot with a ne predicate on boolean field', done => {
```
Comment (stocaaro) [inline packages/datastore/__tests__/DataStore.ts:1781]: This test is very similar. Could some of the setup logic be moved to a beforeFilter?  Would this make the tests easier or harder to follow?

File: packages/datastore/__tests__/DataStore.ts (hunk 4 above)
Comment (svidgen) [reply packages/datastore/__tests__/DataStore.ts:1742]: Yes.

File: packages/datastore/__tests__/DataStore.ts (hunk 7 above)
Comment (svidgen) [reply packages/datastore/__tests__/DataStore.ts:1781]: I'm going to think about this one for followup. If I can templatize these `observeQuery` tests, that would be awesome and make it a bit easier to build more of them up if-needed. I'm not 100% sure how this would affects readability yet. But, I'm struggling to think of how to do this without making things harder to trace through ATM.

File: packages/datastore/__tests__/DataStore.ts (hunk 5 above)
Comment (svidgen) [reply packages/datastore/__tests__/DataStore.ts:1757]: Sprinkled a few comments throughout that hopefully help narrate the test. 🤷

**Review by dpilch — APPROVED** [review]

**Review by stocaaro — APPROVED** [review]

**Review by iartemiev — APPROVED** [review]
